tutorial/practicals/LFRic/single_node/3_sequential/README.md

```diff
@@ -0,0 +1,102 @@
+
+You should see that the loops have not been fused and there should be
+an error output by the transformation. This is because the builtins
+can operate on fields from any function space and fields on different
+function spaces can have different numbers of dofs. Therefore, as
+PSyclone does not know whether the loop bounds of the loops are the
+same it refuses to fuse them.
+
+Let's assume that we know that the fields are all on the same function
+space even though PSyclone does not. In this case we can tell the
+PSyclone transformation that it is safe to fuse. Add same_space=True
+as an argument when the transformation is created:
+
+    ftrans = DynamoLoopFuseTrans(same_space=True)
+
+Rerun psyclone. You should now see that all the loops have been fused
+together.
+
+## Kernel constants
+
+We're now going back to the usual example. There are various sizes
+that are passed to kernels as they are configurable at run-time. For
+example, the number of levels. If the compiler is given these sizes at
+compile time then it could potentially generate better code. Of course
+this means that the kernels are specialised for these particular
+values but codes are often run many times with the same configuration
+(same resolution, etc).
+
+Run psyclone with the kernel_constants.py script:
+
+> psyclone -s ./kernel_constants.py ../code/helmholtz_solver_alg_mod.x90 -oalg /dev/null -opsy psy.f90
+
+On the screen you should see a summary of kernel arguments that have
+been changed.
+
+The transformation is able to provide constant values for the number
+of layers, the order of the problem (which determines how many dofs
+per cell there are and whether the number of quadrature points should
+be constant or not.
+
+The resultant modified kernel code is output in the same directory as
+this file by default.
+
+Feel free to play with the values in the kernel_constants.py script
+and check whether the values change in the modified kernel code as
+expected. Remember that the modified kernel code will not overwrite
+any existing kernel code by default and will be written to a file with
+a different index value.
+
+## Kernel intrinsic
+
+In certain cases it might be useful to replace Fortran instrinsic
+functions in kernels with inline code. In this case we are going to
+replace a matrix multiply call with inline matrix vector code.
+
+First take a look at the file ../code/scaled_matrix_vector_kernel_mod.F90. You should see a matrix multiply contained within it.
+
+Run psyclone with the matvec_opt.py script
+
+> psyclone -s ./matvec_opt.py ../code/helmholtz_solver_alg_mod.x90 -oalg /dev/null -opsy psy.f90
+
+Now take a look at the transformed
+scaled_matrix_vector_kernel_[NUM]_mod.f90 file within this directory,
+where the relevant file should be the one with the largest value of
+[NUM] (as that is the latest transformed file). You should see that
+the matmul has been replaced within this file.
```

Comment:
"...replaced with the equivalent Fortran within..."
I note that the transformed kernel uses a loop variable `i` but that we don't declare `i`. Ditto for j! 'Fortunately' we don't have an `implicit none` either so the code is actually valid!

Reply:
Big oops. That's about the third bug that's been found with the transformation - eek.

Reply:
I blame the reviewer.